iothub/device/src/DirectMethodRequest.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if ENABLE_MODULES_SDK
+namespace Microsoft.Azure.Devices.Client
+{
+    using System;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Parameters to execute a direct method on the device or module
+    /// </summary>
+    public class DirectMethodRequest
+    {
+        internal DirectMethodRequest() { } // @ailn: for serialization only
+
+        /// <summary>
+        /// Creates an instance of DirectMethodRequest type
+        /// </summary>
+        /// <param name="methodName">Method name</param>
+        /// <param name="responseTimeout">Method timeout</param>
+        /// <param name="connectionTimeout">Device connection timeout</param>
+        /// <exception cref="ArgumentException">If <b>methodName</b> is null or whitespace</exception>
+        public DirectMethodRequest(string methodName, TimeSpan responseTimeout, TimeSpan connectionTimeout)
+        {
+            if (string.IsNullOrWhiteSpace(methodName))
+            {
+                throw new ArgumentException("Canot be empty", nameof(methodName));
+            }
+
+            this.MethodName = methodName;
+            this.ResponseTimeout = responseTimeout;
+            this.ConnectionTimeout = connectionTimeout;
+        }
+
+        /// <summary>
+        /// Creates an instance of DirectMethodRequest type with Zero timeout
+        /// </summary>
+        /// <param name="methodName">Method name</param>
+        /// <param name="responseTimeout">Method timeout</param>
+        /// <exception cref="ArgumentException">If <b>methodName</b> is null or whitespace</exception>
+        public DirectMethodRequest(string methodName, TimeSpan responseTimeout)
+            : this(methodName, responseTimeout, TimeSpan.Zero)
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of DirectMethodRequest type with Zero timeout
+        /// </summary>
+        /// <param name="methodName">Method name</param>
+        /// <exception cref="ArgumentException">If <b>methodName</b> is null or whitespace</exception>
+        public DirectMethodRequest(string methodName)
+            : this(methodName, TimeSpan.Zero, TimeSpan.Zero)
```

Comment:
Is TimeSpan.Zero the default? Won't it timeout immediately?

Reply:
When the method is invoked it checks if the timeout is <= 0 it will set a default timeout of 100 seconds. Logic is in HttpTransportHandler and is the same as on ServiceClient.

---
In reply to: [193909920](https://github.com/Azure/azure-iot-sdk-csharp/pull/501#discussion_r193909920) [](ancestors = 193909920)